tests/framework/artifacts.py

```diff
@@ -0,0 +1,191 @@
+        self,
+        bucket
+    ):
+        """Initialize S3 client."""
+        config = botocore.client.Config(signature_version=botocore.UNSIGNED)
+        # pylint: disable=E1101
+        # fixes "E1101: Instance of '' has no 'Bucket' member (no-member)""
+        self.bucket = boto3.resource('s3', config=config).Bucket(bucket)
+
+    def _fetch_artifacts(self,
+                         artifact_root,
+                         artifact_dir,
+                         artifact_ext,
+                         artifact_type,
+                         artifact_class,
+                         keyword=None):
+        artifacts = []
+        prefix = artifact_root + artifact_dir
+        files = self.bucket.objects.filter(Prefix=prefix)
+        for file in files:
+            if (
+                # Filter by extensions.
+                file.key.endswith(artifact_ext)
+                # Filter by userprovided keyword if any.
+                and (keyword is None or keyword in file.key)
+            ):
+                artifacts.append(artifact_class(self.bucket,
+                                                file.key,
+                                                artifact_type=artifact_type))
+        return artifacts
+
+    def microvms(self, keyword=None):
+        """Return microvms artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_MICROVMS,
+            ArtifactCollection.MICROVM_CONFIG_EXTENSION,
+            ArtifactType.MICROVM,
+            Artifact,
+            keyword=keyword
+        )
+
+    def kernels(self, keyword=None):
+        """Return kernel artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_KERNELS,
+            ArtifactCollection.MICROVM_KERNEL_EXTENSION,
+            ArtifactType.KERNEL,
+            Artifact,
+            keyword=keyword
+        )
+
+    def disks(self, keyword=None):
+        """Return disk artifacts for the current arch."""
+        return self._fetch_artifacts(
+            ArtifactCollection.ARTIFACTS_ROOT,
+            ArtifactCollection.ARTIFACTS_DISKS,
+            ArtifactCollection.MICROVM_DISK_EXTENSION,
+            ArtifactType.DISK,
+            DiskArtifact,
+            keyword=keyword
+        )
+
+
+class ArtifactSet:
+    """Manages a set of artifacts with the same type."""
+
+    def __init__(self, artifacts):
+        """Initialize type and artifact array."""
+        assert len(artifacts) > 0
+        self._type = artifacts[0].type
+        self._artifacts = []
+        self.add(artifacts)
+
+    def add(self, artifacts):
+        """Add artifacts to set."""
+        assert len(artifacts) > 0
+        for artifact in artifacts:
+            assert artifact.type == self._type
+            self._artifacts.append(artifact)
+
+    def artifact(self, index):
```

Comment:
Is this used?
Also, just a personal opinion: the code gets hard to follow when there are lot of `assert`s (in non test files). Is it possible to replace them with `if`s and return `None` or some kind of error if those conditions are not met / reduce the number of `assert`s / group them on the same line?

Reply:
I think you are right, this fn was used at some point but forgot to delete it as I was refactoring the code. 

Asserts are important to enforce guarantees so that undefined behavior does not propagate 
to actual test files.

Reply:
`assert`s are a good way of detecting whether an invalid state is detected from which the execution path will not recover.  

Since we don't always want to recover from a faulty test setup, I think that asserts are useful.

Perhaps explaining why they're there would be nice - pytest supports assert messages as shown [here](https://docs.pytest.org/en/latest/assert.html):

```python
assert a % 2 == 0, "value was odd, should be even"
```